### DISCUSSION TITLE
A mental exercise (challenge) regarding UART (?)

### DISCUSSION THREAD

**Discussion by 4yerba** [TG9541/stm8ef#452] [General]
This is not related to stm8ef or stm8 microcontrollers in particular but to  microcontrollers in general. 
I am seeking an informed, ubiased advice on how to proceed further with my UART issue.

Here is the problem I face:
I have a microcontroller that runs a program (a Forth all in one everything) and does it well.
I also have another microcontroller (same, big brand name and type) that apears to be almost the same that is flashed with the same program the reference one is but stubbornly refuses to respond to the UART console input. 
Electrically, both circuits are correct, with proper decoupling capacitors, fed from the same power source, no mixed RX/TX and other pins, pull ups/downs etc. and both micros are perfectly uperational (tested with another image flashed and run).
Inernally, the micros are the same - memory map, same peripherals and so on. All the data matches.

The only difference I have found between these two micros is a JTAG ID, which differs but it's not used anywhere in the sorce code during any stage of the process (compilation, assembly, flashing, running - nowhere).

I have reviewed the datasheets for both of these micros, the user manual, which is the same for both, and even compared the errata sheets for both witch are separate but the same in all relevant parts. All matches up.

Here is the mental challenge: how's that possible ?

The only plausible explanation I find (unconfirmed as of yet) is an existence of an undoccumented feature as the "problematic" micro is branded as a specialty (dedicated) chip that is licensed to run some IP protected library the regular chip is not supposed to run. 
I am totally guessing here but have a feeling that that's why the chips have different JTAG ID that is probably used within the proprietary library source code. 

Any thoughts, suggestions on how to proceed with this puzzle ?

**Comment by TG9541** [comment]
That's an interesting problem. 

The following things come to my mind:
* try a simpler program that just uses communication
* check for RAM initialization issues
* devices can have subtle differences in peripherals configuration
  * the datasheet may not disclose every feature - so variants look equal but a "somewhat OK" initialization here may "not be working" there
* the second device may be defective
* the second device may be counterfeit
  * it's not uncommon that a Chinese manufacturer offers a clone of a device that's mostly compatible which is certainly OK as long as the IP of the original manufacturer is not infringed. Unfortunately there are cases where chip markings get replaced by criminals. Such a chip may work in some applications.

**Reply by 4yerba** [reply]
Thank you fot taking your time and for the suggestions.

The las two ones I have already ruled out. The chip is genuine and the other one I have behaves in exactly same manner.

The subtle differences in peripheral configurations or behaviour I'd expect to be at least mentioned in the reference manual, that is same for the whole range of these chips and undergone numerous revisions and it mentions these differences with regard to other familly members but not to this particular chip so I assumed there is nothing to be mentioned (I may be wrong on this one as I've seen deliberate lies in the datasheets or manuals).

The simple test program that uses just the uart communication I will try next. It's a good idea. The simplest echo that makes use of the physical RX/TX UART pins should be good enough and if that fails I'll dig deeper and check the relevant registers bit by bit (the mundane task I was hoping to avoid but will not hesitate to do it if no other sensible options remain).

I'll report back with the findings. Good or bad.

Thank you

**Comment by VK6TT** [comment]
Hi
If you have a cro make sure the UART pins on the micro are actually changing state. I have been caught by dud rs232 chips before. 
Regards
Richard 

⁣Get BlueMail for Android ​

**Reply by 4yerba** [reply]
Hi,
What's a cro ? 
Good point about the dud usb>uart converters but no, this is not the case. I've thought of that also and used two of them, both verified and perfectly functional elsewhere, different types (not just different pieces of the same). I'll probably resort to a comparative work and torture another chip of the same kind prior to performing step by step debugging. I hate debuggers ;) Sounds stupid but I really do.

**Comment by VK6TT** [comment]
A CRO is an abbreviation for cathode ray oscilloscope. If you don't have one, or access to one, you're doing it the hard way. A logic analyser might help, but that's cumbersome for just confirming the UART is working.

Regards
Richard

⁣Get BlueMail for Android ​

On Oct 12, 2022, 8:03 PM, at 8:03 PM, 4yerba ***@***.***> wrote:
>Hi,
>What's a cro ? 
>Good point about the dud usb>uart converters but no, this is not the
>case. I've thought of that also and used two of them, both verified and
>perfectly functional elsewhere, different types (not just different
>pieces of the same). I'll probably resort to a comparative work and
>torture another chip of the same kind prior to performing step by step
>debugging. I hate debuggers ;) Sounds stupid but I really do.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/TG9541/stm8ef/discussions/452#discussioncomment-3859877
>You are receiving this because you commented.
>
>Message ID:
***@***.***>

**Reply by TG9541** [reply]
> A CRO is an abbreviation for cathode ray oscilloscope. If you don't have one, or access to one, you're doing it the hard way. A logic analyser might help, but that's cumbersome for just confirming the UART is working.

I have an old Fluke Scopemeter - but the cheap USB "8 channel logik analyzer" below (below $8), used with the free [PulseView](https://sigrok.org/wiki/PulseView) turned out to be even more useful when testing communication code.  
[![image](https://user-images.githubusercontent.com/5466977/197352150-4c261c56-e09e-4691-bd69-a243f39b3b85.png)](https://www.aliexpress.com/item/1005002881004277.html)

**Reply by 4yerba** [reply]
A CRO I do not have. I've been toying with the idea of getting one of the Picoscopes' DSO, the 3000 series but just couldn't justify the cost for hobby use. I do have one of these cheap logic analyzers like the one above, however.
 
I will get back to diagnosing the issue probably this very weekend but nonetheless I struggle (and fail so far) to understand how is that possible that seemingly identical chips (sans the JTAG ID which is different) behave differently. Oh, well. Maybe I need to reverse the order  and just check first and  try to understand later. 
Thank you both for the suggestions.

**Comment by sbridger** [comment]
A different JTAG ID might point to a different mask revision of the same chip.
PICs for example have a chip rev #, which is a dead giveaway that the Chinese have relabelled 15 year old chips as 2 years old. (I got PIC16F628A's from 5 different Aliexpress suppliers. All were old mask revs that had been relabelled)
Microchip publishes a separate errata sheet for each part. i.e errata are not in the datasheets.

**Comment by VK6TT** [comment]
I hadn't twigged to that. While I rarely use pics, though I sometimes reprogram a commercial product with flashforth, it's good to know. Given the chip shortages, that manufacturer might have pressed anything into service recently so I'm glad to hear about the Chinese efforts to help 😜
Thanks.

⁣Get BlueMail for Android ​

On Nov 2, 2022, 10:59 AM, at 10:59 AM, sbridger ***@***.***> wrote:
>A different JTAG ID might point to a different mask revision of the
>same chip.
>PICs for example have a chip rev #, which is a dead giveaway that the
>Chinese have relabelled 15 year old chips as 2 years old. (I got
>PIC16F628A's from 5 different Aliexpress suppliers. All were old mask
>revs that had been relabelled)
>Microchip publishes a separate errata sheet for each part. i.e errata
>are not in the datasheets.
>
>
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/TG9541/stm8ef/discussions/452#discussioncomment-4032751
>You are receiving this because you commented.
>
>Message ID:
***@***.***>

**Comment by sbridger** [comment]
@VK6TT BTW, have you made a word to fully sleep, and wake on UART?
I just put some RX antennas down the bank, and they might need a remote switch soon. I want all clocks off in idle.

**Reply by TG9541** [reply]
> @VK6TT BTW, have you made a word to fully sleep, and wake on UART? I just put some RX antennas down the bank, and they might need a remote switch soon. I want all clocks off in idle.

I don't know if I've understood use case and chip (PIC?) correctly - but the [code here](https://gist.github.com/TG9541/61db6e1bdef35d10a7aa02e321d99aa6) might do something similar.

**Reply by sbridger** [reply]
This is indeed what I want to do. I was thinking of using the stm8s003 as I have a couple of the little boards.

**Reply by TG9541** [reply]
That should be possible - the clock settings are a bit different, the WUT has to be used for the background task, and the pin-change interrupt will have to be configured for port D.

**Comment by VK6TT** [comment]
Yes.

I did the following:


$5235 CONSTANT UART_CR2

: Someword

     [ $00 UART_CR2 ]c! \ disables both Rx and Tx

;


However, when trying to remember what I had done I had cause to look at 
the STM8 Reference Manual.

It seems that you could also do the following which could be better from 
a power consumption perspective but I haven't tested this:

$5234 CONSTANT UART_CR1

: Someword

     [ $01 UART_CR1 $05 ]B! \ explicitly also disables UART prescaler

;

**Comment by VK6TT** [comment]
Apologies, I mis-read your message.

No, I have not written a word that sleeps and then wakes on a UART input.

I have slept, then woken on a key-press. I think you'll have to turn off 
the UART, set the pin to input with perhaps a pull-up, then wait for a 
change in the edge. Then wake-up and turn the UART on. I had a button 
connected to PA2, press to ground pin. My approach was to put the STM8 
into Halt ( or sleep) until the key was pressed. Rather than pickup from 
where the code stopped I simply forced a cold start because I had no 
need to restore the state prior to sleeping. I could have put the button 
on the reset line but I chose to use PA2 since I was using button 
presses when the device was awake for user input.


\ on rst EXTI_CR1 is $00 = falling edge and low level all inputs
\ this works since PA2 has pull-up on
: SLEEP \ effectively turn off
    $0 AWU_TBR C! \ reduce power consumption
    [ 1 PA_CR2 _MUT ]B! \ enable ext int.
    [ $8E C, ] \ the HALT instruction
;

\ EXTI0 handler, forces power on reset
:NVM              \ interrupt handler, "headerless" code
    SAVEC
    $80  WWDG_CR C! \ lets force a power on reset
    IRET
;NVM ( xt ) EXTI0 !

**Comment by TG9541** [comment]
@sbridger maybe you should have a look at the link that I put in the reply yesterday - it's quite possible that it does what you need.

**Comment by VK6TT** [comment]
Yes, as always Thomas you're a step ahead of me. Good stuff.

Now I'm left trying to think of a suitable project to apply this to.

**Reply by sbridger** [reply]
Well a remote antenna switch obviously. Another useful variant is a masthead antenna tuner, where the relays/caps are at the mast head, and the swr bridge is back at the radio. It is only relay settings that have to be sent to the tuner.
I'm now trying to work out if the noise on the new antenna is from the antenna, or is being picked up by the UTP cable that brings it up to and through the house. If the UTP is not sufficient, then I have to use a remote antenna switch and a single coax. 
I should have bought a roll of grease filled STP cable instead, but I didn't get it for this purpose.
